chain-signatures/node/src/protocol/message.rs

```diff
@@ -666,175 +721,200 @@ impl MessageReceiver for NodeState {
-            let (from, to, msg, instant) = match outgoing.try_recv() {
+            let (msg, (from, to, timestamp)) = match outgoing.try_recv() {
                 Ok(msg) => msg,
                 Err(TryRecvError::Empty) => {
                     break;
                 }
                 Err(TryRecvError::Disconnected) => {
                     tracing::error!(
-                        "outbox: communication disconnected, no more messages will be received"
+                        "outbox: channel disconnected, no more messages will be received"
                     );
                     break;
                 }
             };
-            self.messages.push_back((from, to, msg, instant));
+            // add it to the outbox and sort it by from and to participant
+            let entry = self.messages.entry((from, to)).or_default();
+            entry.push((msg, timestamp));
+            message_count += 1;
         }
         crate::metrics::MESSAGE_QUEUE_SIZE
             .with_label_values(&[self.account_id.as_str()])
-            .set(self.messages.len() as i64);
+            .set(message_count);
     }
 
     /// Expire messages that have been in the outbox for too long.
     pub fn expire(&mut self, cfg: &ProtocolConfig) {
         // timeout errors are very common for a message expiring, so map them to counts here:
         let mut timeouts = HashMap::<String, usize>::new();
-        self.messages.retain(|(_, to, msg, instant)| {
-            if instant.elapsed() > timeout(msg, cfg) {
-                let counter = timeouts
-                    .entry(format!(
-                        "timeout message={} for node={to:?}",
-                        msg.typename(),
-                    ))
-                    .or_insert(0);
-                *counter += 1;
-                false
-            } else {
-                true
-            }
-        });
+        for ((_from, to), messages) in self.messages.iter_mut() {
+            messages.retain(|(msg, timestamp)| {
+                if timestamp.elapsed() > timeout(msg, cfg) {
+                    let counter = timeouts
+                        .entry(format!(
+                            "timeout message={} for node={to:?}",
+                            msg.typename(),
+                        ))
+                        .or_insert(0);
+                    *counter += 1;
+                    false
+                } else {
+                    true
+                }
+            });
+        }
 
         if !timeouts.is_empty() {
             tracing::warn!(?timeouts, "messages expired");
         }
     }
 
+    /// Compact the messages in the outbox into partitions of at most 256kb.
+    pub fn compact(&mut self) -> HashMap<MessageRoute, Vec<Partition>> {
+        let mut compacted = HashMap::new();
+        for (route, messages) in self.messages.drain() {
+            let entry = compacted.entry(route).or_insert_with(Vec::new);
+            entry.extend(partition_256kb(messages));
```

Comment:
Have you experimented with the optimal size of the partition? Not sure what is the average size of it in practice, but a quick Google search suggested 1-10Mb.